### PR TITLE
Connected UDP Sockets Pool

### PR DESCRIPTION
**Pull request by dv00d00**
Thi pull request modifies PooledUdpTransport class by using a pool of connected udp sockets as discussed in #85 

Benchmarks results:

Before:

|          Method |     Mean |     Error |    StdDev |   Median | Scaled |  Gen 0 | Allocated |
|---------------- |---------:|----------:|----------:|---------:|-------:|-------:|----------:|
| SendWithoutPool | 53.76 us | 1.8980 us | 2.5338 us | 51.65 us |   1.00 | 0.1221 |     440 B |
|    SendWithPool | 14.08 us | 0.0811 us | 0.1189 us | 14.05 us |   0.26 | 0.0610 |     208 B |

After:

|Method |      Mean |     Error |    StdDev | Scaled |  Gen 0 | Allocated |
|--------------------- |----------:|----------:|----------:|-------:|-------:|----------:|
|                 Send | 51.714 us | 0.1015 us | 0.0847 us |   1.00 | 0.1221 |     440 B |
|         SendWithPool |  8.890 us | 0.1292 us | 0.1209 us |   0.17 |      - |      40 B |
| SendWithPoolSwitcher | 11.839 us | 0.0425 us | 0.0398 us |   0.23 | 0.0153 |      67 B |

`SendWithPoolSwitcher` bench tests edge case when endpoint changes every milisecond

### CODE REVIEW

**Comment by dv00d00** [comment]
I was able to reproduce test failures on my mac, created a ticket in corefx (https://github.com/dotnet/corefx/issues/31206)

**Comment by martincostello** [comment]
The failed test could just maybe be a manifestation of differences in OS behaviours a bit like how we have to have this code:

https://github.com/justeat/JustEat.StatsD/blob/554633f110b6e9292a5d69055c171e9a0ebc32da/src/JustEat.StatsD/UdpTransport.cs#L52-L56

File: src/JustEat.StatsD/PooledUdpTransport.cs [hunk 1]
```diff
@@ -10,13 +11,71 @@ namespace JustEat.StatsD
...
+            public ConnectedPool(IPEndPoint ipEndPoint)
+            {
+                IpEndPoint = ipEndPoint;
+
+                Pool = new SimpleObjectPool<Socket>(
+                    Environment.ProcessorCount,
+                    pool =>
+                    {
+                        var socket = UdpTransport.CreateSocket();
+                        socket.Connect(ipEndPoint);
```
Comment (martincostello) [inline src/JustEat.StatsD/PooledUdpTransport.cs]: Need to `try-catch` around `Connect()` and dispose in the catch block, as otherwise if it fails we'll leak a socket.

File: src/JustEat.StatsD/PooledUdpTransport.cs [hunk 2]
```diff
@@ -10,13 +11,71 @@ namespace JustEat.StatsD
     /// </summary>
     public sealed class PooledUdpTransport : IStatsDTransport, IDisposable
     {
-        private readonly SimpleObjectPool<Socket> _socketPool
-            = new SimpleObjectPool<Socket>(
-                Environment.ProcessorCount,
-                pool => UdpTransport.CreateSocket());
+        private sealed class ConnectedPool : IDisposable
```
Comment (martincostello) [inline src/JustEat.StatsD/PooledUdpTransport.cs]: This private class isn't trivial, so can you make it internal and put it in its own file please?

File: src/JustEat.StatsD/PooledUdpTransport.cs [hunk 3]
```diff
@@ -40,66 +99,58 @@ public void Send(string metric)
...
             }
 
-            _socketPool.Push(socket);
+            pool.Pool.Push(socket);
         }
-
-        /// <summary>
-        /// Finalizes an instance of the <see cref="PooledUdpTransport"/> class.
-        /// </summary>
-        ~PooledUdpTransport()
```
Comment (martincostello) [inline src/JustEat.StatsD/PooledUdpTransport.cs]: The `IDisposable` implementation should be retained.

File: src/JustEat.StatsD/PooledUdpTransport.cs [hunk 4]
```diff
@@ -40,66 +99,58 @@ public void Send(string metric)
 
             var bytes = Encoding.UTF8.GetBytes(metric);
             var endpoint = _endpointSource.GetEndpoint();
+            var pool = GetPool(endpoint);
+            var socket = pool.Pool.PopOrCreate();
```
Comment (martincostello) [inline src/JustEat.StatsD/PooledUdpTransport.cs]: Is there a better name we can have for this, as `pool.Pool` doesn't scan well?

File: src/JustEat.StatsD/PooledUdpTransport.cs [hunk 5]
```diff
@@ -40,66 +99,58 @@ public void Send(string metric)
...
 
-        /// <summary>
-        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
-        /// </summary>
-        /// <param name="disposing">
-        /// <see langword="true" /> to release both managed and unmanaged resources;
-        /// <see langword="false" /> to release only unmanaged resources.
-        /// </param>
-        private void Dispose(bool disposing)
+        private readonly object _lock = new object();
```
Comment (martincostello) [inline src/JustEat.StatsD/PooledUdpTransport.cs]: Field declarations should live together.

File: src/JustEat.StatsD/PooledUdpTransport.cs [hunk 6]
```diff
@@ -40,66 +99,58 @@ public void Send(string metric)
...
-                        throw;
-                    }
+                    _pool.Dispose();
+                    _pool = new ConnectedPool(endPoint);
+                    return _pool;
                 }
-
-                _disposed = true;
             }
+
```
Comment (martincostello) [inline src/JustEat.StatsD/PooledUdpTransport.cs]: Blank line not needed.

File: src/JustEat.StatsD/PooledUdpTransport.cs [hunk 7]
```diff
@@ -40,66 +99,58 @@ public void Send(string metric)
...
-                catch (ObjectDisposedException)
+            }
+
+            if (endPoint.Equals(_pool.IpEndPoint))
+            {
+                return _pool;
+            }
+            else
+            {
+                lock (_lock)
```
Comment (martincostello) [inline src/JustEat.StatsD/PooledUdpTransport.cs]: I don't think this lock is a good idea, as it's effectively putting a lock around network calls to `Connect()` in the constructor of the pool.

File: src/JustEat.StatsD/PooledUdpTransport.cs [hunk 8]
```diff
@@ -40,66 +99,58 @@ public void Send(string metric)
...
                 {
-                    // If the pool has already been disposed by the finalizer, ignore the exception
-                    if (disposing)
-                    {
-                        throw;
-                    }
+                    _pool.Dispose();
+                    _pool = new ConnectedPool(endPoint);
```
Comment (martincostello) [inline src/JustEat.StatsD/PooledUdpTransport.cs]: If any of the connections in the constructor throw, then subsequent calls will return the disposed-of pool, putting everything in a bad state. Something like an optimistic `Interlocked.CompareExchange()` might be better, as that removes the need for a lock and does not corrupt state if the initialization fails.

Note this doesn't check whether the pool that was created before the thread created its pool is connected to the same endpoint.

```csharp
var oldPool = _pool;
var newPool = new ConnectedPool(endPoint);

try
{
    if (Interlocked.CompareExchange(ref _pool, oldPool, newPool) == oldPool)
    {
        We successfully exchanged the pools
        oldPool.Dispose();
    }
    else
    {
        // Something already replaced the pool
        newPool.Dispose();
    }
}
catch (Exception)
{
    newPool.Dispose();
}
```

File: src/JustEat.StatsD/PooledUdpTransport.cs (hunk 3 above)
Comment (slang25) [reply src/JustEat.StatsD/PooledUdpTransport.cs]: It's a sealed class, so there would be no point to this dispose pattern.

**Comment by dv00d00** [comment]
I've fixed Unix tests by binding loopback ports to a UDP socket as seen in UdpListeners.cs

File: src/JustEat.StatsD.Tests/WhenUsingPooledUdpTransport.cs [hunk 9]
```diff
@@ -1,15 +1,26 @@
...
+        }
+
         [Fact]
-        public static void AMetricCanBeSentWithoutAnExceptionBeingThrown()
+        public void AMetricCanBeSentWithoutAnExceptionBeingThrown()
         {
             // Arrange
-            var endPointSource = EndpointLookups.EndpointParser.MakeEndPointSource("127.0.0.1", 8125, null);
+            var endPointSource = EndpointParser.MakeEndPointSource(
+                UdpListeners.EndpointA.Address.ToString(),
```
Comment (martincostello) [inline src/JustEat.StatsD.Tests/WhenUsingPooledUdpTransport.cs]: This makes me wonder whether we should have an overload that accepts an `IPEndPoint`.

File: src/JustEat.StatsD/ConnectedSocketPool.cs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+                {
+                    var socket = UdpTransport.CreateSocket();
+                    try
+                    {
+                        socket.Connect(ipEndPoint);
+                        return socket;
+                    }
+                    catch
+                    {
+                        socket?.Dispose();
```
Comment (martincostello) [inline src/JustEat.StatsD/ConnectedSocketPool.cs]: `socket` won't ever be `null` here.

File: src/JustEat.StatsD/ConnectedSocketPool.cs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+                        throw;
+                    }
+                });
+        }
+
+        public IPEndPoint IpEndPoint { get; }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (martincostello) [inline src/JustEat.StatsD/ConnectedSocketPool.cs]: Can you before & after the perf for with an without having to add these attributes? I'm curious as to what difference they make.

Comment (dv00d00) [reply src/JustEat.StatsD/ConnectedSocketPool.cs]: Attribute did not make any difference because method was already inlined by compiler, but this is the difference between inlined

 |Method |     Mean |     Error |    StdDev | Allocated |
|------- |---------:|----------:|----------:|----------:|
|    Act | 79.10 ns | 0.7256 ns | 0.6433 ns |       0 B |

and MethodImplOptions.NoInlining methods

| Method |     Mean |    Error |   StdDev | Allocated |
|------- |---------:|---------:|---------:|----------:|
|    Act | 84.92 ns | 1.681 ns | 2.186 ns |       0 B |

Comment (dv00d00) [reply src/JustEat.StatsD/ConnectedSocketPool.cs]: Code for benchmark was:

```csharp

    [MemoryDiagnoser]
    public class InliningBenchmark
    {
        private static readonly ConnectedSocketPool _csp =
            new ConnectedSocketPool(new IPEndPoint(IPAddress.Loopback, 1234));

        [Benchmark]
        public void Act()
        {
            var socket = _csp.PopOrCreate();
            _csp.Push(socket);
        }
    }

```

File: src/JustEat.StatsD.Tests/WhenUsingPooledUdpTransport.cs (hunk 9 above)
Comment (martincostello) [reply src/JustEat.StatsD.Tests/WhenUsingPooledUdpTransport.cs]: What do you think @AnthonySteele ?

Comment (AnthonySteele) [reply src/JustEat.StatsD.Tests/WhenUsingPooledUdpTransport.cs]: So the fix for the tests on mac and linux was purely in the test setup?

**Comment by martincostello** [comment]
Thank you for your contribution @dv00d00! 🥇

File: src/JustEat.StatsD.Tests/WhenUsingPooledUdpTransport.cs (hunk 9 above)
Comment (martincostello) [reply src/JustEat.StatsD.Tests/WhenUsingPooledUdpTransport.cs]: #92
